common/src/main/java/com/wynntils/core/managers/ManagerRegistry.java

```diff
@@ -57,105 +56,88 @@ public static void init() {
         }
 
-        MODEL_DEPENDENCIES.putIfAbsent(dependency, new ArrayList<>());
+        List<ModelDependant> managerDependencies =
+                MODEL_DEPENDENCIES.computeIfAbsent(manager, (k) -> new ArrayList<>());
 
-        MODEL_DEPENDENCIES.get(dependency).remove(dependant);
+        managerDependencies.remove(dependant);
 
-        updateManagerState(dependency);
+        return !managerDependencies.isEmpty() || tryDisableManager(manager);
     }
 
-    public static void removeAllFeatureDependency(Feature dependant) {
-        for (Class<? extends Manager> manager : MODEL_DEPENDENCIES.keySet()) {
-            boolean removed = MODEL_DEPENDENCIES.get(manager).remove(dependant);
+    public static void removeAllFeatureDependency(ModelDependant dependant) {
+        for (Class<? extends Model> model : dependant.getModelDependencies()) {
 
-            if (removed) {
-                updateManagerState(manager);
+            List<ModelDependant> managerDependencies = MODEL_DEPENDENCIES.get(model);
+
+            if (managerDependencies != null && managerDependencies.remove(dependant)) {
+                tryDisableManager(model);
             }
         }
     }
 
-    private static void updateManagerState(Class<? extends Manager> manager) {
-        List<Feature> dependencies = MODEL_DEPENDENCIES.get(manager);
-        List<Function<?>> functionDependencies = FUNCTION_MODEL_DEPENDENCIES.get(manager);
-
-        boolean hasDependencies = (dependencies != null && !dependencies.isEmpty())
-                || (functionDependencies != null && !functionDependencies.isEmpty());
-
-        if (ENABLED_MANAGERS.contains(manager)) {
-            if (!hasDependencies) {
-                WynntilsMod.unregisterEventListener(manager);
-
-                ENABLED_MANAGERS.remove(manager);
-
-                tryDisableManager(manager);
-            }
-        } else {
-            if (hasDependencies) {
-                WynntilsMod.registerEventListener(manager);
+    private static boolean tryInitManager(Class<? extends Manager> manager) {
+        // add if not already present
+        if (!ENABLED_MANAGERS.add(manager)) return false;
 
-                ENABLED_MANAGERS.add(manager);
-
-                tryInitManager(manager);
-            }
-        }
-    }
+        WynntilsMod.registerEventListener(manager);
 
-    private static void tryInitManager(Class<? extends Manager> manager) {
         try {
             MethodUtils.invokeExactStaticMethod(manager, "init");
         } catch (IllegalAccessException | NoSuchMethodException e) {
             WynntilsMod.warn("Misconfigured init() on manager " + manager, e);
-            assert false;
+            return false;
         } catch (InvocationTargetException e) {
-            WynntilsMod.error("Exception during init of manager " + manager, e.getTargetException());
+            WynntilsMod.warn("Exception during init of manager " + manager, e.getTargetException());
```

Comment:
And why are you downgrading this error to a warning? :-(

@Incompleteusern, you **need** to learn more discipline when coding. You are constantly changing a lot of side stuff that is not relevant to the issue you are trying to solve. Often these changes are controversial or misguided. When reviewing your PRs, I feel I spend more time just on the look for this kind of "what the heck is happening here" than trying to understand your *intent* and if the way you have written the patch is a proper way to solve the stated goal.

Please, please try to keep to the rule "one issue, one PR". In the PR, start by explaining what is the issue you are trying to resolve.

At times, it might be necessary to have a "collect a bunch of small fixes"-PR, since it might be too much overhead to have a single PR to e.g. demote this error to a warning. But then such a PR should not do anything else than being a bunch of minor cleanups, and it needs to be reviewed differently.

Fixing a typo or an incorrect comment that is not relevant to the PR *might* be okay to include, but even that is pushing it. (I will accept it though.) But not this kinds of changes, which seem innocuous but radically changes behavior. 

I've been trying to guide you before in trying not to mix issues, but apparently I need to be much more clear. This is not a good way of software development. It makes reviewing hard. It increases the risk that bugs are slipping into the code. And it makes it hard to understand the history of the code base. And focusing on one issue at the time forces the developer to fully engage in that particular problem.

There just is no upside to this kinds of sloppy PRs.